File: 3445-lexicographically-minimum-string-after-removing-stars/lexicographically-minimum-string-after-removing-stars.cpp
class Solution {
public:
    string clearStars(string s) {
        priority_queue<pair<char,int>,vector<pair<char,int>>, greater<pair<char,int>>> pq;
        vector<int> ind;
        for(int i=0;i<s.size();i++){
            if(s[i]!='*'){
                pq.push({s[i],-i});
            }
            else{
                ind.push_back(i);
                ind.push_back(-pq.top().second);pq.pop();
            }
        }
        sort(ind.begin(),ind.end());
        string ans="";int j=0;
        for(int i=0;i<s.size();i++){
            if(j<ind.size() && i==ind[j]){j++;continue;}
            else ans+=s[i];
        }
        return ans;
    }
};

File: 3445-lexicographically-minimum-string-after-removing-stars/lexicographically-minimum-string-after-removing-stars.java
class Solution {
    public String clearStars(String s) {
        int n = s.length();
        PriorityQueue<Character> pq = new PriorityQueue<>();
        Map<Character,ArrayList<Integer>> mp=new HashMap<>();
        boolean[] keep = new boolean[n];
        Arrays.fill(keep,true);
        for(int i=0;i<n;i++){
            char c = s.charAt(i);
            if(c == '*'){
                char temp = pq.poll();
                keep[i] = false;
                keep[mp.get(temp).remove(mp.get(temp).size()-1)] = false;
            }
            else{
                pq.add(c);
                mp.putIfAbsent(c,new ArrayList<Integer>());
                mp.get(c).add(i);
            }
        }

        StringBuilder ans = new StringBuilder();
        for (int i = 0; i < n; i++){
            if (keep[i]) {
                ans.append(s.charAt(i));
            }
        }
        return ans.toString();
        
    }
}